molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditingRepositoryDecorator.java

```diff
@@ -0,0 +1,340 @@
+                audit(ENTITY_READ, ENTITY_ID, entity.getIdValue());
+                return true;
+              });
+    } else {
+      return delegate().findAll(ids);
+    }
+  }
+
+  @Override
+  public Stream<Entity> findAll(Stream<Object> ids, Fetch fetch) {
+    if (isRunByUser() && isNonSystemEntityType()) {
+      return delegate()
+          .findAll(ids, fetch)
+          .filter(
+              entity -> {
+                audit(ENTITY_READ, ENTITY_ID, entity.getIdValue());
+                return true;
+              });
+    } else {
+      return delegate().findAll(ids, fetch);
+    }
+  }
+
+  @Override
+  public void deleteAll() {
+    delegate().deleteAll();
+
+    if (isRunByUser()) {
+      audit(ALL_ENTITIES_DELETED);
+    }
+  }
+
+  @Override
+  public void add(Entity entity) {
+    delegate().add(entity);
+
+    if (isRunByUser()) {
+      audit(ENTITY_CREATED, ENTITY_ID, entity.getIdValue());
+    }
+  }
+
+  @Override
+  public Integer add(Stream<Entity> entities) {
+    if (isRunByUser()) {
+      entities =
+          entities.filter(
+              entity -> {
+                audit(ENTITY_CREATED, ENTITY_ID, entity.getIdValue());
+                return true;
+              });
+    }
+
+    return delegate().add(entities);
+  }
+
+  @Override
+  public void update(Entity entity) {
+    delegate().update(entity);
+
+    if (isRunByUser()) {
+      audit(ENTITY_UPDATED, ENTITY_ID, entity.getIdValue());
+    }
+  }
+
+  @Override
+  public void update(Stream<Entity> entities) {
+    if (isRunByUser()) {
+      entities =
+          entities.filter(
+              entity -> {
+                audit(ENTITY_UPDATED, ENTITY_ID, entity.getIdValue());
+                return true;
+              });
+    }
+
+    delegate().update(entities);
+  }
+
+  @Override
+  public void delete(Entity entity) {
+    delegate().delete(entity);
+
+    if (isRunByUser()) {
+      audit(ENTITY_DELETED, ENTITY_ID, entity.getIdValue());
+    }
+  }
+
+  @Override
+  public void delete(Stream<Entity> entities) {
+    if (isRunByUser()) {
+      entities =
+          entities.filter(
+              entity -> {
+                audit(ENTITY_DELETED, ENTITY_ID, entity.getIdValue());
+                return true;
+              });
+    }
+
+    delegate().delete(entities);
+  }
+
+  @Override
+  public void deleteAll(Stream<Object> ids) {
+    if (isRunByUser()) {
+      ids =
+          ids.filter(
+              id -> {
+                audit(ENTITY_DELETED, ENTITY_ID, id);
+                return true;
+              });
+    }
+
+    delegate().deleteAll(ids);
+  }
+
+  @Override
+  public void deleteById(Object id) {
+    delegate().deleteById(id);
+
+    if (isRunByUser()) {
+      audit(ENTITY_DELETED, ENTITY_ID, id);
+    }
+  }
+
+  private void audit(String type) {
+    var data = createDataMap();
+    auditEventPublisher.publish(getUsername(), type, data);
+  }
+
+  private void audit(String type, String k1, Object v1) {
```

Comment:
key, value probably slightly more meaningful parameter names

Reply:
Ah yes, I was copying the `Map.of(v1, v2, k1, v2)` style here but I had multiple methods then :)